Add convenience functions

rhendric opened the pull request:
`tailRec2` and `tailRec3` complete the `tailRec` and `tailRecM` matrix,
and `loop2` and `loop3` are new conveniences for both `tailRec2/3` and
`tailRecM2/3` that better abstract the consumer away from the details of
the `a`, `b`, `c` record labels we use in those functions.

**Description of the change**

Submitting this for discussion in preparation for doing something about purescript/purescript#4289. All of these new functions are optional for that proposal, but before we go hard on promoting `purescript-tailrec` as the best way to ensure that one's stack-hungry functions execute safely, I think we should address these rough edges, and standardizing on them will make it possible to optimize them out.

---

**Checklist:**

- [x] Added the change to the changelog's "Unreleased" section with a reference to this PR (e.g. "- Made a change (#0000)")
- [ ] Linked any existing issues or proposals that this pull request should close
- [ ] Updated or added relevant documentation
- [x] Added a test for the contribution (if applicable)

JordanMartinez commented on src/Control/Monad/Rec/Class.purs:
```diff
@@ -93,6 +97,14 @@ tailRec f = go <<< f
   go (Loop a) = go (f a)
   go (Done b) = b
 
+-- | Create a pure tail-recursive function of two arguments
```
Could you add docs that refer to `loop2` here? Otherwise, it's not immediately clear that one would know to use `loop2` in tandem with `tailRec2`

JordanMartinez commented on src/Control/Monad/Rec/Class.purs:
```diff
@@ -93,6 +97,14 @@ tailRec f = go <<< f
+tailRec2 :: forall a b c. (a -> b -> Step { a :: a, b :: b } c) -> a -> b -> c
+tailRec2 f a b = tailRec (\o -> f o.a o.b) { a, b }
+
+-- | Create a pure tail-recursive function of three arguments
```
Similarly here for `loop3`?

rhendric committed:
Add convenience functions

`tailRec2` and `tailRec3` complete the `tailRec` and `tailRecM` matrix,
and `loop2` and `loop3` are new conveniences for both `tailRec2/3` and
`tailRecM2/3` that better abstract the consumer away from the details of
the `a`, `b`, `c` record labels we use in those functions.

JordanMartinez reviewed (approved):
LGTM!

JordanMartinez commented:
This was released in `v6.1.0`.
